checker/src/org/checkerframework/checker/index/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -184,62 +188,7 @@ private void addUpperBoundTypeFromValueType(
 
     @Override
     protected DependentTypesHelper createDependentTypesHelper() {
-        return new DependentTypesHelper(this) {
-            @Override
-            protected String standardizeString(
-                    final String expression,
-                    FlowExpressionContext context,
-                    TreePath localScope,
-                    boolean useLocalScope) {
-                if (DependentTypesError.isExpressionError(expression)) {
-                    return expression;
-                }
-                if (indexOf(expression, '-', '+', 0) == -1) {
-                    return super.standardizeString(expression, context, localScope, useLocalScope);
-                }
-
-                OffsetEquation equation = OffsetEquation.createOffsetFromJavaExpression(expression);
-                if (equation.hasError()) {
-                    return equation.getError();
-                }
-                try {
-                    equation.standardizeAndViewpointAdaptExpressions(
-                            context, localScope, useLocalScope);
-                } catch (FlowExpressionParseUtil.FlowExpressionParseException e) {
-                    return new DependentTypesError(expression, e).toString();
-                }
-
-                return equation.toString();
-            }
-
-            private int indexOf(String expression, char a, char b, int index) {
-                int aIndex = expression.indexOf(a, index);
-                int bIndex = expression.indexOf(b, index);
-                if (aIndex == -1) {
-                    return bIndex;
-                } else if (bIndex == -1) {
-                    return aIndex;
-                } else {
-                    return Math.min(aIndex, bIndex);
-                }
-            }
-
-            @Override
-            public TreeAnnotator createDependentTypesTreeAnnotator(AnnotatedTypeFactory factory) {
-                return new DependentTypesTreeAnnotator(factory, this) {
-                    @Override
-                    public Void visitMemberSelect(MemberSelectTree tree, AnnotatedTypeMirror type) {
-                        // UpperBoundTreeAnnotator changes the type of array.length to @LTEL
-                        // ("array"). If the DependentTypesTreeAnnotator tries to viewpoint
-                        // adapt it based on the declaration of length; it will fail.
-                        if (TreeUtils.isArrayLengthAccess(tree)) {
-                            return null;
-                        }
-                        return super.visitMemberSelect(tree, type);
-                    }
-                };
-            }
-        };
+        return new OffsetDependentTypesHelper(this);
```

Comment:
Abstracting this into a new class is a fine choice.  Why did you create a new file for it rather than making it a nested class in this file?

Reply:
`OffsetDependentTypesHelper` does not use any internals of `UpperBoundAnnotatedTypeFactory`, it is not used exclusively by it and there is no logical relationship between these classes. Therefore I don't see any reason for nesting it.
Also, I put it in the common `org.checkerframework.checker.index` package, because it is used by both the Upper bound and IndexOf checkers. However, the dependence on `OffsetEquation` might be a reason to put it in the `org.checkerframework.checker.index.upperbound` package.